app/src/behaviors/behavior_mod_tap.c

```diff
@@ -46,22 +158,92 @@ static int behavior_mod_tap_init(struct device *dev)
 static int on_keymap_binding_pressed(struct device *dev, u32_t position, u32_t mods, u32_t keycode)
 {
   struct behavior_mod_tap_data *data = dev->driver_data;
-  LOG_DBG("mods: %d, keycode: %d", mods, keycode);
-  WRITE_BIT(data->pending_press_positions, position, true);
-  return ZMK_EVENT_RAISE(create_modifiers_state_changed(mods, true));
+  LOG_DBG("mods: %d, keycode: 0x%02X", mods, keycode);
+  for (int i = 0; i < ZMK_BHV_MOD_TAP_MAX_HELD; i++) {
+    if (data->active_mod_taps[i].mods != 0) {
+      continue;
+    }
+
+    zmk_mod_flags active_mods = behavior_mod_tap_active_mods(data);
+
+    data->active_mod_taps[i].active_mods = active_mods;
+    data->active_mod_taps[i].mods = mods;
+    data->active_mod_taps[i].keycode = keycode;
+    data->active_mod_taps[i].pending = true;
+
+    return 0;
+  }
+
+  LOG_WRN("Failed to record mod-tap activation, at maximum concurrent mod-tap activations");
+
+  return -ENOMEM;
 }
 
 static int on_keymap_binding_released(struct device *dev, u32_t position, u32_t mods, u32_t keycode)
 {
   struct behavior_mod_tap_data *data = dev->driver_data;
   LOG_DBG("mods: %d, keycode: %d", mods, keycode);
   
-  ZMK_EVENT_RAISE(create_modifiers_state_changed(mods, false));
-  k_msleep(10); // TODO: Better approach than k_msleep to avoid USB send failures? Retries in the USB endpoint layer?
-  if (data->pending_press_positions & BIT(position)) {
-    ZMK_EVENT_RAISE(create_keycode_state_changed(USAGE_KEYPAD, keycode, true));
-    k_msleep(10);
-    ZMK_EVENT_RAISE(create_keycode_state_changed(USAGE_KEYPAD, keycode, false));
+  for (int i = 0; i < ZMK_BHV_MOD_TAP_MAX_HELD; i++) {
+    struct active_mod_tap_item *item = &data->active_mod_taps[i];
+    if (item->mods == mods && item->keycode == keycode) {
+      if (item->pending) {
+        LOG_DBG("Sending un-triggered mod-tap for keycode: 0x%02X", keycode);
+
+        if (item->active_mods) {
+          LOG_DBG("Registering recorded active mods captured when mod-tap initially activated: 0x%02X", item->active_mods);
+          behavior_mod_tap_update_active_mods_state(data, item->active_mods);
+          zmk_hid_register_mods(item->active_mods);
+        }
+
+        struct keycode_state_changed *key_press = create_keycode_state_changed(USAGE_KEYPAD, item->keycode, true);
+        ZMK_EVENT_RAISE_AFTER(key_press, behavior_mod_tap);
+        k_msleep(10);
+
+        for (int j = 0; j < ZMK_BHV_MOD_TAP_MAX_PENDING_KC; j++) {
+          if (data->captured_keycode_events[j].event == NULL) {
+            continue;
+          }
+
+          struct keycode_state_changed *ev = data->captured_keycode_events[j].event;
+          data->captured_keycode_events[i].event = NULL;
+          data->captured_keycode_events[i].active_mods = 0;
+          LOG_DBG("Re-sending latched key press for usage page 0x%02X keycode 0x%02X state %s", ev->usage_page, ev->keycode, (ev->state ? "pressed" : "released"));
+          ZMK_EVENT_RELEASE(ev);
+          k_msleep(10);
```

Comment:
Yeah, that's a band-aid to avoid the USB send failures more often. It's not perfect, since we can have key press/release which are super close together that still cause issues.